src/util/graph.h

```diff
@@ -584,6 +602,92 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
   return result;
 }
 
+/// Run recursive depth-limited search on the graph, starting
+/// from multiple source nodes, to find the nodes reachable within n steps.
+/// This function initialises the search.
+/// \param src The node to start the search from.
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+/// \param limit  limit on steps
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const node_indext &src,
+  std::size_t &limit,
+  bool forwards) const
+{
+  std::vector<node_indext> start_vector(1, src);
+  return depth_limited_search(start_vector, limit, forwards);
+}
+
+/// Run recursive depth-limited search on the graph, starting
+/// from multiple source nodes, to find the nodes reachable within n steps.
+/// This function initialises the search.
+/// \param src The nodes to start the search from.
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+/// \param limit  limit on steps
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
+  std::size_t &limit,
+  bool forwards) const
+{
+  std::vector<bool> visited(nodes.size(), false);
+
+  for(const auto &node : src)
+    visited[node] = true;
+
+  return depth_limited_search(src, limit, visited, forwards);
+}
+
+/// Run recursive depth-limited search on the graph, starting
+// from multiple source nodes, to find the nodes reachable within n steps
+/// \param src The nodes to start the search from.
+/// \param limit  limit on steps
+/// \param visited vector of booleans indicating whether a node has been visited
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
```

Comment:
This would be much simpler if it worked over single sources rather lists-of-sources

Reply:
Yes it would. I wrote the function to work with a vector of sources to be consistent with `grapht::get_reachable`, which has a function for a vector of sources and one for a single source. I don't know when you would want to do reachability from multiple sources, but given that is an option, it seems nicely symmetric to also have an option to do depth-limited reachability from multiple sources.

Reply:
Sure, but the multi-source one can just be a wrapper, like
```
std::set<irep_idt> depth_limited_search(
  const std::vector<node_indext> &srcs, size_t depth)
{
  std::set<irep_idt> result;
  for(const node_indext src : srcs)
    depth_limited_search(src, result, depth);
  return result;
}
```

Reply:
Argh, true, sorry. Would breadth-first search do the trick?

Reply:
(In case you've trashed the old version of this but want to get back to it -- `git reflog` lets you see where a branch used to point to and recover the old "lost" commits)

Reply:
Re BFS: I think you get in trouble whenever you use an implementation that uses recursion, because you need to communicate the depth of when you have seen a node across invocations.

Reply:
I used BFS last time I implemented this,  and, if I recall correctly, @smowton reviewed it and said that I should switch it to DFS in order to use more efficient containers. 

(The way to do it with BFS is to mark the end of each frontier with some specific node index, I think I used numeric_limits::max)

Reply:
I wasn't gunning for DFS in particular, I was trying to eliminate the copying of the results or frontier vector at each recursive call. However I've clearly made far more mess than is justified here, so I'll butt out.

Reply:
The BFS implementation didn't involve recursion and copying, just a for loop. But it didn't use vectors, it used a stack. 

Either way, I'd prefer it if we just stuck with one version now so that I can redo all my changes on top of that. it's easier for me to revert to the DFS version, since the BFS version was based on the old call graph class so I would pretty much have to rewrite it.

Reply:
Oh I see, I thought you were referring to the review a few days ago, not a really old one